File: algo-rust/src/hash/hash.rs
#![allow(dead_code)]
use std::collections::HashMap;

struct FindSumPairs {
    nums1: Vec<i32>,
    nums2: Vec<i32>,
    map: HashMap<i32, i32>,
}

impl FindSumPairs {
    fn new(nums1: Vec<i32>, nums2: Vec<i32>) -> Self {
        let mut map = HashMap::new();

        for num in nums2.iter() {
            *map.entry(*num).or_insert(0) += 1;
        }

        FindSumPairs { nums1, nums2, map }
    }

    fn add(&mut self, index: i32, val: i32) {
        *self.map.entry(self.nums2[index as usize]).or_insert(1) -= 1;
        self.nums2[index as usize] += val;
        *self.map.entry(self.nums2[index as usize]).or_insert(0) += 1;
    }

    fn count(&self, tot: i32) -> i32 {
        let mut res = 0;

        for a in self.nums1.iter() {
            if let Some(v) = self.map.get(&(tot - a)) {
                res += v;
            }
        }
        res
    }
}

struct Solution {}

impl Solution {
    // leetcode 2325
    pub fn decode_message(key: String, message: String) -> String {
        let mut map = HashMap::new();
        let mut index: u8 = 97;
        for c in key.chars() {
            if !c.is_alphabetic() || map.get(&c).is_some() {
                continue;
            }

            map.insert(c, index as char);
            index += 1;
        }

        let mut result = String::new();
        for c in message.chars() {
            if c.is_alphabetic() && map.contains_key(&c) {
                result.push(map[&c]);
            } else {
                result.push(c);
            }
        }

        result
    }
    // leetcode 1282
    pub fn group_the_people(group_sizes: Vec<i32>) -> Vec<Vec<i32>> {
        let mut result = Vec::new();
        let mut group_map = HashMap::new();

        for (i, size) in group_sizes.iter().enumerate() {
            let group = group_map.entry(size).or_insert(vec![]);

            group.push(i as i32);

            if group.len() >= *size as usize {
                if let Some((_, group)) = group_map.remove_entry(size) {
                    result.push(group);
                }
            }
        }

        result
    }

    // Longest Consecutive Sequence
    pub fn longest_consecutive(nums: Vec<i32>) -> i32 {
        let (mut map, mut ans) = (HashMap::new(), 0);

        for n in nums {
            if map.contains_key(&n) {
                continue;
            }

            map.insert(n, 1);

            let (mut left, mut right, mut length) = (n, n, 1);
            if let Some(v) = map.get(&(n - 1)) {
                left = n - v;
                length += map.get(&left).cloned().unwrap_or(0);
            };

            if let Some(v) = map.get(&(n + 1)) {
                right = n + v;
                length += map.get(&right).cloned().unwrap_or(0);
            }

            map.insert(left, length);
            map.insert(right, length);
            ans = std::cmp::max(ans, length);
        }
        ans
    }

    // Count Number of Pairs With Absolute Difference K
    pub fn count_k_difference(nums: Vec<i32>, k: i32) -> i32 {
        let (mut res, mut map) = (0, HashMap::new());
        for num in nums {
            if let Some(v) = map.get(&(num - k)) {
                res += v;
            }

            if let Some(v) = map.get(&(num + k)) {
                res += v;
            }

            *map.entry(num).or_insert(0) += 1;
        }

        res
    }
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn group_the_people() {
        let group_sizes = vec![3, 3, 3, 3, 3, 1, 3];
        let want = vec![vec![0, 1, 2], vec![5], vec![3, 4, 6]];

        assert_eq!(want, Solution::group_the_people(group_sizes));
    }
    #[test]
    fn longest_consecutive() {
        let nums = vec![100, 4, 200, 1, 3, 2];
        let want = 4;

        assert_eq!(want, Solution::longest_consecutive(nums));
    }
}

File: algo-go/hash/hash.go
package hash

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

// smallerNumbersThanCurrent 1365. 有多少小于当前数字的数字
// 计数排序
func smallerNumbersThanCurrent(nums []int) []int {
	count := make(map[int]int)

	for i := 0; i < len(nums); i++ {
		count[nums[i]]++
	}

	for i := 1; i <= 100; i++ {
		count[i] += count[i-1]
	}

	result := make([]int, len(nums))

	for i := 0; i < len(nums); i++ {
		if nums[i] == 0 {
			result[i] = 0
		} else {
			result[i] = count[nums[i]-1]
		}
	}

	return result
}

// distributeCandies leetcode 575. 分糖果
func distributeCandies(candyType []int) int {
	mp := map[int]struct{}{}

	for _, n := range candyType {
		mp[n] = struct{}{}
	}

	return min(len(candyType)/2, len(mp))
}

func min(a, b int) int {
	if a > b {
		return b
	}

	return a
}

// longestConsecutive 128. 最长连续序列
func longestConsecutive(nums []int) int {
	set := make(map[int]struct{})

	for _, n := range nums {
		set[n] = struct{}{}
	}

	ans := 0

	for n := range set {
		if _, ok := set[n-1]; !ok {
			cur := n
			curLen := 1

			for {
				_, ok := set[cur+1]
				if !ok {
					break
				}

				cur++
				curLen++
			}

			ans = max(ans, curLen)
		}
	}
	return ans
}

// romanToInt leetcode 13. 罗马数字转整数
func romanToInt(s string) int {
	m := map[string]int{
		"I": 1,
		"V": 5,
		"X": 10,
		"L": 50,
		"C": 100,
		"D": 500,
		"M": 1000,
	}
	res := 0

	n := len(s)
	for i := 0; i < n; i++ {
		var k1 string
		var k2 string
		k1 = s[i : i+1]
		if i+2 <= n {
			k2 = s[i+1 : i+2]
		}

		if m[k1] < m[k2] {
			res += m[k2] - m[k1]
			i++
		} else {
			res += m[k1]
		}
	}
	return res
}
